### PR TITLE
Security part 2

### PR DESCRIPTION
**Pull request by ihorromaniuk**

### CODE REVIEW

File: src/main/java/core/basesyntax/bookstore/repository/user/UserRepository.java [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
 
 import core.basesyntax.bookstore.model.User;
+import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    @Query("from User u left join fetch u.roles where u.email = :email")
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/repository/user/UserRepository.java]: it's better to use
```suggestion
    @EntityGraph(attributePaths = "roles")
```

File: src/main/java/core/basesyntax/bookstore/security/UserDetailsServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsServiceImpl implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email).orElseThrow(()
+                -> new UsernameNotFoundException("Can't find user by email. Email: " + email));
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/security/UserDetailsServiceImpl.java]:
```suggestion
                -> new EntityNotFoundException("Can't find user by email. Email: " + email));
```

File: src/main/java/core/basesyntax/bookstore/security/UserDetailsServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsServiceImpl implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/security/UserDetailsServiceImpl.java]:
```suggestion
    public UserDetails loadUserByUsername(String email) {
```

File: src/main/java/core/basesyntax/bookstore/exception/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -31,6 +33,21 @@ public ResponseEntity<ExceptionDto> handleRegistrationException(Exception ex) {
         return new ResponseEntity<>(exceptionDto, status);
     }
 
+    @ExceptionHandler(UsernameNotFoundException.class)
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/exception/GlobalExceptionHandler.java]: remove and handle EntityNotFoundException

File: src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -4,24 +4,41 @@
...
-        return userMapper.toDto(repository.save(userMapper.toModel(requestDto)));
+        User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.password()));
+        Role userRole = roleRepository.getFirstByRole(Role.RoleName.USER);
+        user.setRoles(Set.of(userRole));
+        return userMapper.toDto(userRepository.save(user));
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/UserServiceImpl.java]: remove

File: src/main/java/core/basesyntax/bookstore/service/UserService.java [hunk 6]
```diff
@@ -2,7 +2,11 @@
 
 import core.basesyntax.bookstore.dto.user.UserDto;
 import core.basesyntax.bookstore.dto.user.UserRegistrationRequestDto;
+import core.basesyntax.bookstore.model.User;
+import java.util.Optional;
 
 public interface UserService {
     UserDto save(UserRegistrationRequestDto requestDto);
+
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/UserService.java]: remove

File: src/main/java/core/basesyntax/bookstore/BookstoreApplication.java [hunk 7]
```diff
@@ -1,17 +1,15 @@
...
-import core.basesyntax.bookstore.service.BookService;
-import org.springframework.beans.factory.annotation.Autowired;
+import static org.springframework.data.web.config.EnableSpringDataWebSupport.PageSerializationMode.VIA_DTO;
+
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.data.web.config.EnableSpringDataWebSupport;
 
 @SpringBootApplication
+@EnableSpringDataWebSupport(pageSerializationMode = VIA_DTO)
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/BookstoreApplication.java:10]: Why do you need this annotation?

File: src/main/java/core/basesyntax/bookstore/model/Role.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/Role.java:15]:
```suggestion
@Entity
@Getter
@Setter
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/core/basesyntax/bookstore/BookstoreApplication.java (hunk 7 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/BookstoreApplication.java:10]: image

File: src/main/java/core/basesyntax/bookstore/model/User.java [hunk 9]
```diff
@@ -31,4 +38,45 @@ public class User {
...
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/model/User.java]: make it soft-deleted entity and use 
```suggestion
    public boolean isEnabled() {
        return !isDeleted;
    }
```

File: src/main/java/core/basesyntax/bookstore/repository/user/RoleRepository.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.bookstore.repository.user;
+
+import core.basesyntax.bookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role getFirstByRole(Role.RoleName role);
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/bookstore/repository/user/RoleRepository.java]: better return Optional 
```suggestion
    Optional<Role> findByName(Role.RoleName role);
```

File: src/main/resources/db/changelog/changes/02-create-role-table.yaml [hunk 11]
```diff
@@ -0,0 +1,32 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-role-table
+      author: Ihor Romaniuk
+      objectQuotingStrategy: QUOTE_ONLY_RESERVED_WORDS
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-role-table.yaml:5]: why do you need it? the same for other scripts

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few comments ;)

File: src/main/resources/db/changelog/changes/02-create-role-table.yaml (hunk 11 above)
Comment (ihorromaniuk) [reply src/main/resources/db/changelog/changes/02-create-role-table.yaml:5]: it's auto generated

Comment (JJJazl) [reply src/main/resources/db/changelog/changes/02-create-role-table.yaml:5]: You write your scripts by yourself, it mustn't be auto generated)

File: src/main/resources/db/changelog/changes/02-create-role-table.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                  type: VARCHAR(255)
+            tableName: roles
+  - changeSet:
+      id: add-unique-constraint-name
+      author: Ihor Romaniuk
+      objectQuotingStrategy: QUOTE_ONLY_RESERVED_WORDS
+      changes:
+        - addUniqueConstraint:
+            columnNames: name
+            constraintName: uc_roles_role
```
Comment (JJJazl) [inline src/main/resources/db/changelog/changes/02-create-role-table.yaml:30]: Let's adhere naming convention for contraints like "unique", "foreign key", "primary key". Check everywhere

**Review by JJJazl — APPROVED** [review]

**Pull request merged**